### ISSUE TITLE
Authorize.net payment gateway issue

### ISSUE THREAD

**Issue by sohelmagento** [magento/magento2#23915]
**Preconditions**

Magento CE 2.2.4
PHP 7.1.30
MySQL 5.6.30

**Steps to reproduce**
1. Follow all these instruction provide by magento and authorize.net for integration.
2. Set to live payment and also setup backend for live payment.
3. Fill all required field like API ID, Transaction key, Signature key and both URL.
4. Add the signature key module to bypass MD5 key.
5. Increased transaction timeout.
6. When payment with authorize.net after fill all information it giving "sorry, but something went wrong. please contact the seller." error.
7. On test and live mode. Both time give same error.
8. After few second it's going to success page or press OK to go success page.
9. We was discus with authorize.net team and verify my log file but they are told us it's not a our response error it's a third party module error.
10. We added payment.log and debug.log file for check response and bug. Please provide solution ASAP.

Response payment.log
[payment.log](https://github.com/magento/magento2/files/3437889/payment.log)

Magento debug.log
[debug.log](https://github.com/magento/magento2/files/3437890/debug.log)

**Expected Results**
Go to success page without any attention error popup.

**Actual Results**
It's going to success page but before go to success it's showing this error "sorry, but something went wrong. please contact the seller." and after that go to success page. If we press OK it's go to success page and if we not press OK so after few seconds it's goes to success page.

![authorize-net-issue](https://user-images.githubusercontent.com/53362794/61990597-0729b180-b061-11e9-8b40-ab696a2f5e27.png)

**Comment by irajneeshgupta** [comment]
@sohelmagento 

since you are using `Magento CE 2.2.4` , i assume that you have applied patch for `md5 to sha-256`
and then generate  fresh production keys from authorize.net. 
Try the following for production :
test mode : `No`
gateway url : `https://secure.authorize.net/gateway/transact.dll`
txn detail url : `https://api2.authorize.net/xml/v1/request.api`

In my case test , this issue arise while using test mode `yes` with test keys.
and test urls are different and did not auto applied while test mode yes. 
Try to use above configuration.

**Comment by sohelmagento** [comment]
@irajneeshgupta 
Thanks your response,

Yes i have applied patch for md5 to sha-256 and i am using the same configration that you have mentioned but i am still having this same error. No matter i try in test and production mode.

**Comment by irajneeshgupta** [comment]
@sohelmagento 
This error generally occurs when no txn details return from authorize.net or a timeout occured.
You can view the response in `var/log/payments.log` 
It should look like this :

`
[2019-07-19 15:50:39] main.DEBUG: array (
  'response' => 
  array (
    'x_response_code' => '1',
    'x_response_reason_code' => '1',
    'x_response_reason_text' => 'This transaction has been approved.',
    'x_avs_code' => 'Y',
    'x_auth_code' => '56869P',
    'x_trans_id' => '41474336050',
    'x_method' => 'CC',
    'x_card_type' => 'MasterCard',
    'x_account_number' => 'XXXX0300',
    'x_first_name' => 'Luke',
    'x_last_name' => 'Myers',
    'x_company' => '',
    'x_address' => '1420 The Strand',
    'x_city' => 'Manhattan Beach',
    'x_state' => 'California',
    'x_zip' => '90266',
    'x_country' => 'US',
    'x_phone' => '3107365853',
    'x_fax' => '',
    'x_email' => 'test@gmail.com',
    'x_invoice_num' => '1000027987',
    'x_description' => '',
    'x_type' => 'auth_capture',
    'x_cust_id' => '',
    'x_ship_to_first_name' => 'Ethan',
    'x_ship_to_last_name' => 'Weitz',
    'x_ship_to_company' => '',
    'x_ship_to_address' => '201 Ocean Reef Drive',
    'x_ship_to_city' => 'Key Largo',
    'x_ship_to_state' => 'Florida',
    'x_ship_to_zip' => '33037-5273',
    'x_ship_to_country' => 'US',
    'x_amount' => '363.90',
    'x_tax' => '0.00',
    'x_duty' => '0.00',
    'x_freight' => '0.00',
    'x_tax_exempt' => 'FALSE',
    'x_po_num' => '',
    'x_MD5_Hash' => '',
    'x_SHA2_Hash' => '997C4C9A94ABCEFD3128123AE513576799C057E75A3FBE3879B1B8BADFF888BA4F52E70ACC19DDF5CD8E69253567C623F98DE00A6FABD74A5436D3021C33AXXX',
    'x_cvv2_resp_code' => 'M',
    'x_cavv_response' => '',
    'x_test_request' => 'false',
    'controller_action_name' => 'directpost_payment',
    'is_secure' => '1',
  ),
) ` 

In my case , x_trans_id was 0 in case of test setup and due to 0 ,condition gets failed in validations.
You should get the error in exception.log or if your magento is in dev mode , you can view the error in console.

**Comment by sohelmagento** [comment]
@irajneeshgupta,
Thanks your response,

We discus about this error with authorize.net support team they told us your response code is correct it's third party module exception please contact them. For transaction id 0 in test mode we change some code for solve this issue. 
```
vendor/magento/module-authorizenet/Model/Directpost.php
if (!$this->getResponse()->getXTransId()) {
replace with
if (!$this->getResponse()->getXTestRequest() && !$this->getResponse()->getXTransId()) {
```
But still showing this error.

**Comment by sdzhepa** [comment]
Hello @sohelmagento  and @engcom-Charlie 

Could you please confirm that issue can be reproduced by provided steps on **clear Magento Open Sorce 2.3 installation**?

In case is it related only to `2.2.x` version we should close this issue because Magento no longer accepts pull requests related to the 2.2

**Comment by engcom-Charlie** [comment]
Hi @sohelmagento and @sdzhepa 
I can't reproduce this issue on fresh Magento 2.3-develop.
@sohelmagento can you check it on 2.3-develop instance? I will request instance for you or you can just write "@magento give me 2.3-develop instance"
Thanks!

**Comment by magento-engcom-team** [comment]
Hi @engcom-Charlie. Thank you for your request. I'm working on Magento 2.3-develop instance for you

**Comment by magento-engcom-team** [comment]
Hi @engcom-Charlie, here is your [Magento instance](https://i-23915-2-3-develop.instances.magento-community.engineering).
Admin access: [https://i-23915-2-3-develop.instances.magento-community.engineering/admin](https://i-23915-2-3-develop.instances.magento-community.engineering/admin)
Login: `admin` Password: `123123q`
Instance will be terminated in up to 3 hours.

**Comment by irajneeshgupta** [comment]
@sdzhepa @sohelmagento

It is 2.2.x issue and can be resolved if you can upgrade magento to 2.3.x series. You should close this issue since it is already fixed in latest versions.

**Comment by sohelmagento** [comment]
Thanks @irajneeshgupta @sdzhepa and @engcom-Charlie ,

Yes it's resolved in magento 2.3.x .
